### ISSUE TITLE
Sending metrics to any Azure resrouce from the SDK

### ISSUE THREAD

**Issue by macrogreg** [microsoft/ApplicationInsights-dotnet#783]
Hi folks,
 
Azure is standardizing an ingestion API for metrics. 

There are a few key differences between that API and the current Application Insights API that matter here:

- The new API can accept metrics about any Azure resource, not only Application Insights components.
- The new API identifies resources by their ARM resource ID, not their IKey.
- The new API requires authentication (via AAD tokens).
- The new API is regional. Metrics must be explicitly sent to the endpoint located in the same region as the target resource.
 
We need to support sending metrics about any Azure resource to the new ingestion endpoint from the SDK.

We had some discussions about the high level approach with @SergeyKanzhelev  and  @karolz-ms . The goal here is to discuss this with a wider community and to document our plans. Also, note that this is high-level for now. As the discussion evolves, we will add more precise details in comments to this issue.
 
*Key assumption*:
The vision is to evolve the AI SDK to the “One Azure Telemetry SDK”. In the process of this we will likely add a central, general way to send any kind of telemetry to any kind of Azure resources. We are not solving this general problem for now. The goal at hand is to merely send metric telemetry to the right place with an explicit understanding that addressing and authentication approach we use may evolve over time.

*Solution approach*:
 
- We will use the “sink” functionality available in the SDK pipeline. We will create a sink for sending metrics to the new endpoint. The sink will be configured in the normal manner (like any other App Insights config today).
 
- Authentication information will be part of the sink configuration (e.g. certificate thumbprints necessary to obtain AAD tokens).
 
- In v1, the endpoint address will be an explicit part of the sink configuration. I.e. the user must configure the endpoint for the right region. In subsequent versions we will use ARM discovery APIs to automatically determine the correct region. 
 
- To address metrics we will use the existing iKey field: 

  - If the string contained in the iKey field of a metric telemetry item has a specific well-defined prefix, the SDK will assume that the sub-string after that prefix is an ARM resource ID and not an instrumentation key. It will then use the aforementioned sink to send the telemetry item.

  - Otherwise the item will be send through the “normal” Application Insights channel.
 
- Setting the iKey field for metric telemetry items can happen using any of the existing approaches. For example: 

  - Set the `InstrumentationKey` field on the `TelemetryClient`, and use `MetricAggregationScope.TelemetryClient` value in `GetMetric(..)` methods;

  - Use a metric dimension with the name defined in the `MetricDimensionNames.TelemetryContext.InstrumentationKey` constant and the value of the ARM resource ID with the aforementioned prefix (or the iKey).

  - Use `InstrumentationKey` on the application `TelemetryConfiguration` instance.

  - Issue: The first 2 approaches are verbose: they require a setting for every metric. This is fine when many different resources must be addressed, but cumbersome when all metrics target the same resources.
On the other hand, setting `InstrumentationKey` on `TelemetryConfiguration` will prevent non-metric telemetry from working. We need to address this in v2. One possibility is to address this is to add a property like `InstrumentationKey` to the `MetricManager`.

**Comment by SergeyKanzhelev** [comment]
I'd suggest starting by implementing sink in the separate NuGet as ServerTelemetryChannel is implemented. It is likely will be minimal overhead for creating it. It is tempting to move it inside the Base SDK directly, but I'd suggest not to rush with it.

**Comment by macrogreg** [comment]
I am personally fine with a separate NuGet. But then we need to understand two things:

- Decide whether they will take some kind of dependency on each other, or whether Web or another NuGet will take a dependency on both. Basically what "things" and how many different "things" does a user need to install in order to get this functionality?

- What is the perspective of the business org on this? 

Either way, whatever NuGets we go with, it will be a key part of the SDK ecosystem, so discussing this Issue here is appropriate. Concerns?

**Comment by SergeyKanzhelev** [comment]
Can you please start with separate NuGet than. For now let make people add this new channel explicitly so no need to reference it from base SDK.

One Azure Telemetry SDK effort and implications lead me to believe this approach may be a bit better until we cleared out all details.

**Comment by macrogreg** [comment]
> [ @karolz-ms  ] One possible alternative would be to have a new standard telemetry property ("AzureResourceId" or similar). New endpoint would be used if InstrumentationKey was not set and AzureResourceId was.

I agree and something of this kind was my initial proposal. In fact, I suggested it would be an object property rather than a string, and the resource-id would be a string property on that object. Polymorphism on that object would allow us to be flexible in evolving the "addressing" requirements over time. We would also be able to potentially group addressing and authentication info.

However, @SergeyKanzhelev has pointed out that he preferred the aforementioned iKey-reusing-approach to start with. I will let him provide the detailed reasons. My high level impression was that he preferred to not introduce new concepts before the space of sending _any_ telemetry to _any_ Azure resources has been approached holistically.

**Comment by macrogreg** [comment]
> [ @SergeyKanzhelev ] Can you please start with separate NuGet than.

That works for me. :)

> [ @SergeyKanzhelev ] For now let make people add this new channel explicitly so no need to reference it from base SDK.

For me that works too, but I'd like to make sure that this also works for the business owners/PMs before we consider this point closed. Are you on board?

**Comment by Dmitry-Matveev** [comment]
>>Issue: The first 2 approaches are verbose: they require a setting for every metric. This is fine when many different resources must be addressed, but cumbersome when all metrics target the same resources.
On the other hand, setting InstrumentationKey on TelemetryConfiguration will prevent non-metric telemetry from working. We need to address this in v2. One possibility is to address this is to add a property like InstrumentationKey to the MetricManager.

It essentially means for any application with auto-collected telemetry (that also happens to use metrics), Azure Resource ID would need to be specified on each metric to avoid losing said auto-collected telemetry. It's only partilly (if at all) related to the fact we're re-using IKey field here - this seems to be the question of scoping "the second channel" to the right entity. Is there anything that may provide a single configuraiton point for a group of related metrics to simplify configuration experience or we are OK as is in V1 from scenario perspective?

**Comment by macrogreg** [comment]
@Dmitry-Matveev , I completely agree. The reason why it is acceptable in v1 is that the majority of applications will be sending telemetry to a single resource only. Supporting multiple resources is a requirement, but we believe that it is an edge case, at least initially.
Despite that, I very much agree with you on that this is still an annoying limitation. In fact, in the context of @SergeyKanzhelev 's recommendation to ship v1 in a separate NuGet, I had an idea of how we could solve this my introducing a small number of new APIs in the separate NuGet only. let me summarize this in a separate comment later today.

**Comment by macrogreg** [comment]
Hi folks,

I have some thoughts on how to architect the metrics submission to the new endpoint within the SDK.
I wonder about your feedback..

We already established that this will be in a different NuGet to start with.
Let us consider how metrics data flows through the SDK today:
 
![currentarchitecture01](https://user-images.githubusercontent.com/20691812/39076503-695dee5a-44c1-11e8-87af-c131aede242c.JPG)

Because the Azure metrics endpoint is different, the in the diagram layers “_Endpoint Client_”, “_Breeze Sink_” and “_Breeze Channel_” cannot be reused. In addition, the layer “_Telemetry Pipeline_” may be reused with limited benefit. Also, the `TelemetryConfiguration` class, that combines the configuration for some of the aforementioned components and their execution does not fully apply to the new metric channel.

I propose that we create a new type of client, say `MetricsClient`. An instance of such client will represent exactly one Azure resource that can receive metrics. There will also be a `MetricsClientConfiguration` class that can be used to configure one or more instances of `MetricsClient`. In v1, configuration will be in code only, in the future we can support config via `ApplicationInsight.config` too.

To send metrics to Breeze, `TelemetryClient `will be used just like it is today. To send metrics to the new endpoint, `MetricsClient `will be used.

Today, metrics are emitted by writing 
```
TelemetryClient telemetryClient = new TelemetryClient(..);
Metric metric = telemetryClient.GetMetric(..);
metric.TrackValue(..);
```

The new client will work in the same way:
```
MetricsClient metricsClient = MetricsClient.ForResource(“Aruze Resource Id”);
Metric metric = metricsClient.GetMetric(..);
metric.TrackValue(..);
```

The existing `MetricIndentifier` class will get a new field to identify the Azure resource.
The `MetricsClient` will sent a new kind of telemetry, `MetricAggregateTelemetry : ITelemetry`, via the normal pipeline.
We make this a new class rather than reusing the existing `MetricTelemetry` class to make sure that existing telemetry processors ignore it and just pass it through. Items of this class will be ignored by the App Insights sink and will be picked up by the Geneva sink.
Summary:

![proposedarchitecture](https://user-images.githubusercontent.com/20691812/39076322-35c11c80-44c0-11e8-8240-f1cf534e96a0.JPG)

In a nutshell, that’s it.
Of course, there are API details to work out.
I wonder what the team thinks about it. Especially, I would welcome feedback about how we can simplify this, while keeping the overall idea that we reuse the processing pipeline of the AI SDK, while decoupling form its channel and singleton config.

Thanks!

**Comment by macrogreg** [comment]
We could also avid creating a new `MetricsClient `class. Instead we could have an alternative way of creating a `TelemetryClient`, e.g.

`TelemetryClient client = TelemetryClient.ForResource("resource id");`

Instances created in that way would only be meant for `client.getMetric(..)` and `TrackXxx(..)` would fail. In the future, as we send more kinds of of telemetry to any kind of resource, we would make `TrackXxx(..)` work accordingly.

However, this static method cannot be implemented in a different NuGet.

**Comment by macrogreg** [comment]
Another way to keep things simple in v1 (relative to the `MetricsCient` proposal above) is to leave participating in the App Insights telemetry pipeline for v2 and to send data right away:

![proposedarchitecture02](https://user-images.githubusercontent.com/20691812/39076954-1bc44880-44c4-11e8-97d4-4e079e84fd4c.JPG)

**Comment by Dmitry-Matveev** [comment]
I cannot say I'm fully onboard with `MetricsClient` approach:
- In interim it will produce two clients: one talking to the new endpoint (`MetricsClient`), one talking to the old endpoint (`TelemetryClient`);
- In the longer term, we may need to make TelemetryClient compatible with new endpoint as well, essentially producing two clients that sends to the same endpoint where only one of them has `getMetric().TrackValue()` API.

Static method `ForResource()` actually looks promising with the exception of the exceptions here and there and allows for future migration of whole `TelemetryClient` into the new space with deprecation of older creation methods if necessary.

Can you please clarification on the third approach where you skip Telemetry Pipeline? What will be implemented in a new Nuget if sink and channel are gone with this approach? Is there a scenario we might lose by choosing this way over the others? I'd expect so, otherwise there would be no reason to have pipeline at all :)

**Comment by SergeyKanzhelev** [comment]
`MetricsClient` may be a good shim that abstracts setting pipeline.

We need somehow generic solution on setting pipeline for specific `TelemetryClient`. It will help implementing QuickPulse as a sink and collect performance counters every second that only intended to this sink. 

There are definitely more scenarios that will benefit from ability to set pipeline per `TelemetryClient` as oppose to for entire `TelemetryConfiguration`.

I know we are under time pressure now. So if you feel you need this `MetricsClient` helper shim - you can look into implementing it in that separate NuGet.

**Comment by macrogreg** [comment]
@Dmitry-Matveev , good points. Thank you.

> Can you please clarification on the third approach where you skip Telemetry Pipeline? What will be implemented in a new Nuget if sink and channel are gone with this approach? Is there a scenario we might lose by choosing this way over the others? I'd expect so, otherwise there would be no reason to have pipeline at all :)

Essentially, the channel must be different anyway and cannot be re-used. The simplification offered by the 3rd option is that we do not use the TelemetryInitializers/-Processors pipeline either, and loose that functionality in V1 (we can always re-add it later). In that approach, only metrics aggregation functionality is shared. Note - that simplification works with both - with the `MetricsClient` and with the `TelemetryClient.ForResource` approaches above.

@SergeyKanzhelev, also great insights. Note that the current implementation of QuickPulse and metrics does not treat QP as a Sink. Instead, the aggregation subsystem and QP interface directly. This is because they need to communicate 2 ways to set up filters and what metrics are even being watched live.

So we have two options:
a) `MetricsClient` that abstracts the authenticated endpoint and only the authenticated endpoint.
b) `TelemetryClient.ForResource` that constructs a special kind of TelemetryClient instance that can only track metrics and throws for other telemetry items (for now). In this option, the static factory method ForResource(..) would need to be declared in the base SDK. 

I am fine with both, but you two guys seem to have conflicting opinions on this. What shall we go with?

**Comment by SergeyKanzhelev** [comment]
I support @Dmitry-Matveev on this. Specially-constructed `TelemetryClient` is more aligned with other long-term requirements. So `TelemetryClient.ForResource` it is

**Comment by LouisStAmour** [comment]
I was looking briefly at what it would take to add metrics support to https://github.com/open-telemetry/opentelemetry-collector-contrib using the newer API, but I was stuck on this whole "resource ID" concept and how it might apply to metrics from non-Azure resources? It appears at first glance that the "legacy" way of using an iKey is the better approach for this kind of generic metrics reporting? 

The alternative is you'd have to tell people to look up the resource ID for the insights group itself, and wouldn't the data end up in effectively the same place, but now instead of knowing one ID, you need up pass up to five pieces of data, including a subscription ID, a region, resource group, resource type and a resource ID? https://davidjenei.com/post/azure-monitor-telegraf/

By comparison, the new log workspaces API seems to be very similar to the previous way it was done, but if I'm reading this right, the keys and values are optionally less specific to code? Why not add optional values to metrics reporting to include resources?

The more I look into this, the more I wonder why these three different endpoints were created with very different, and seemingly incompatible UIs for each. Yes, I can understand that different methods of data collection would require different APIs for optimization, but outside of API standardization efforts such as OpenTelemetry or different agent collection methods such as OpenTelemetry Collectors to perform tail sampling, the only other optimization I can think of would be to use a zero copy binary protocol, perhaps in some kind of application-specific way.

Introducing the flexibility to assign metrics to a resource is perhaps nice, but not if the cost is that metrics behave completely differently from logs and traces, and in a way that's actually less flexible? Just because a resource is in a region, should I send metrics to only that region as well?

Looking at Go and Python, the two languages I've been working with recently, it's interesting to see the many different approaches:

The old way (via batched exports of all three to `https://dc.services.visualstudio.com/v2/track`)
- Logs: https://github.com/census-instrumentation/opencensus-python/blob/08c34fc271421dbed8baad8e62113d30049c5463/contrib/opencensus-ext-azure/opencensus/ext/azure/log_exporter/__init__.py#L167
- Metrics: https://github.com/census-instrumentation/opencensus-python/blob/08c34fc271421dbed8baad8e62113d30049c5463/contrib/opencensus-ext-azure/opencensus/ext/azure/metrics_exporter/__init__.py#L76
- Traces: https://github.com/census-instrumentation/opencensus-python/blob/08c34fc271421dbed8baad8e62113d30049c5463/contrib/opencensus-ext-azure/opencensus/ext/azure/trace_exporter/__init__.py#L59

The "new" way:
- Logs: https://github.com/Azure/oms-log-analytics-firehose-nozzle/blob/14c3bf5cb0ae1eb375aca80f6e967b8210250c5b/client/oms_client.go
- Metrics: https://github.com/influxdata/telegraf/blob/2799302142d3ebf8967107a8dd45431326895954/plugins/outputs/azure_monitor/azure_monitor.go#L309

I do see the note here about log-based metrics vs pre-aggregated metrics -- https://docs.microsoft.com/en-us/azure/azure-monitor/app/pre-aggregated-metrics-log-metrics#pre-aggregated-metrics -- but it doesn't exactly make sense to me. If metrics are reported using a different "envelope" anyway when sent to App Insights, why wouldn't the App Insights telemetry client choose whether to report traces when sampling, but continue to pre-aggregate basic metrics (such as pageview counts) regardless of whether a request was sampled or not? The easy answer to me is that you'd report using a standard metric name and prefer using the standard metric name if it exists, falling back only if the client is older and doesn't report the metric to a log-based metric instead?

**Comment by LouisStAmour** [comment]
After a second day of research, it sounds like https://docs.microsoft.com/en-us/azure/azure-monitor/app/create-workspace-resource is exactly what I'm looking for, and I'll use Application Insights (old batched API) to send all the data. Makes more sense for now, it's the closest thing there is for an Azure all-in-one monitoring solution. I've no idea why it's so hard to figure this out, why it's not easier or more obvious...

**Comment by cijothomas** [comment]
@LouisStAmour Depending on what you are trying to achieve, I can try to help.. 
This issue was opened a long back, and there was zero work done to make it happen. (I am not sure if there will be any work on this in this repo at all.)

The OpenTelemetry exporter (without OTel Collector) will be hosted in Azure SDK repo - https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/monitor
